File: gestion-formasion/src/app/models/filiere.ts
import { Formation } from "./formation";
import { Module } from "./module";
import { Personne } from "./personne";

export class Filiere {
    id: number;
    modules: Module[];
    stagiaires: Personne[];
    formation: Formation;
    libelle:string;
    dateDebut:Date;
    dateFin:Date;

    constructor(id: number, modules: Module[], stagiaires: Personne[], formation: Formation,dateDebut:Date,dateFin:Date) {
        this.id = id;
        this.modules = modules;
        this.stagiaires = stagiaires;
        this.formation = formation;
        this.dateDebut=dateDebut;
        this.dateFin=dateFin;
        this.libelle = this.id+"_"+this.formation.libelle;
    }
}

File: formasionAPI/src/main/java/fr/formation/formasion/model/Filiere.java
package fr.formation.formasion.model;

import java.time.LocalDate;
import java.util.List;
import java.util.Objects;

import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.SequenceGenerator;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@SequenceGenerator(name = "filiere_gen", sequenceName = "filiere_seq", initialValue = 100, allocationSize = 1)
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class Filiere {

	@Id
	@GeneratedValue(generator = "filiere_gen")
	private Integer id;

	@OneToMany(mappedBy = "filiere")
	@JsonIgnoreProperties("filiere")
	private List<Module> modules;

	@OneToMany(mappedBy = "filiere")
	@JsonIgnoreProperties("filiere")
	private List<Personne> stagiaires;

	@ManyToOne
	@JoinColumn(name = "formation_libelle")
	@JsonIgnoreProperties({"filieres","stagiaires"})
	private Formation formation;
	
	public Filiere(Integer id) {
		this.id=id;
	}

	public String getLibelle() {
		return this.id+"_"+this.formation.getLibelle();
	}
	
	public LocalDate getDateDebut() {
		LocalDate min = null;
		for(Module m : modules) {
			if(Objects.isNull(min) || min.isAfter(m.getDateDebut())) min=m.getDateDebut();
		}
		return min;
	}
	
	public LocalDate getDateFin() {
		LocalDate max = null;
		for(Module m : modules) {
			if(Objects.isNull(max) || max.isBefore(m.getDateFin())) max=m.getDateFin();
		}
		return max;
	}

}
